application/cms/dimension_service.py

```diff
@@ -159,6 +189,10 @@ def update_dimension(dimension, data):
                     chart_options[key] = "[None]"
             data["chart_2_source_data"]["chartOptions"] = chart_options
             dimension.chart_2_source_data = data.get("chart_2_source_data")
+            if data["use_custom"] is False:
+                self.__set_chart_dimension_classification_through_builder(dimension, data)
```

Comment:
Same goes for the `__set_chart_dimension...` and `__set_table_dimension...` methods.  Suggestions welcome as to how we refactor these to be one thing.

Reply:
@TheDoubleK how about something like:

```python
def update_dimension(self, dimension, data, update_classification=False):


  if data.get("chart_2_source_data"):

    chart_or_table = Chart.get_by_id(dimension.chart_id) or Chart()
    dimension.chart_id = chart_or_table.id

  elif data.get("table_2_source_data"):

    chart_or_table = Table.get_by_id(dimension.table_id) or Table()
    dimension.table_id = chart_or_table.id

  
  self.__set_dimension_classification(self, chart_or_table, classification_link)    

  db.session.add(dimension)
  db.session.commit()



def __set_dimension_classification(self, chart_or_table, classification_link):

    chart_or_table.classification_id = classification_link.classification_id
    chart_or_table.includes_parents = classification_link.includes_parents
    chart_or_table.includes_all = classification_link.includes_all
    chart_or_table.includes_unknown = classification_link.includes_unknown

    db.session.add(chart_or_table)
    db.session.flush()
```

Reply:
@TheDoubleK actually, perhaps a better alternative is to refactor this to the model mixin?

```python
def update_dimension(self, dimension, data, update_classification=False):

  if data.get("chart_2_source_data"):

    chart_or_table = Chart.get_by_id(dimension.chart_id) or Chart()
    dimension.chart_id = chart_or_table.id

  elif data.get("table_2_source_data"):

    chart_or_table = Table.get_by_id(dimension.table_id) or Table()
    dimension.table_id = chart_or_table.id

  chart_or_table.set_classification(self, classification_link)    

  db.session.add(dimension)
  db.session.commit()


class ChartAndTableMixin(object):

  def set_classification(self, classification_link):

    self.classification_id = classification_link.classification_id
    self.includes_parents = classification_link.includes_parents
    self.includes_all = classification_link.includes_all
    self.includes_unknown = classification_link.includes_unknown

    db.session.add(self)
    db.session.flush()
```

Reply:
On reflection (i.e. I've just had enough now)...
I think we might be better off leaving this refactor until when we move the chart and table data out into the dimension_chart and dimension_table tables.  And then it will make sense to have a method that updates everything on the chart/table in the model mixin.